### PR TITLE
Add CONTRIBUTING for SV405CC support

### PR DESCRIPTION
**Pull request by jctk**
Hi, @thx8411 
I have added SV405CC support and some improvements to indi_sv305_ccd.
Could you review the code and merge if there are no problems?

The changes are as follows:

1. Add SV405CC support
2. FIXED: mutex conflict
- Mutex was defined as Static, so when more than one SVBONY CMOS camera was connected, one mutex would conflict.
3. IMPROVED: OFFSET support on Ekos Camera module.
- Changed the name of the input field from CCD_DOFFSET to CCD_OFFSET to enable the OFFSET field in the Capture Module of Ekos.

Note
If you have ever connected to the SV405CC with a driver that does not include this modification, an incorrect configuration file has been created in ~/.indi.
If there is an incorrect configuration file, please delete it.

### CODE REVIEW

**Review by thx8411 — COMMENTED** [review]
Hi jctk,

Thanks a lot, great Job !

I just received my camera yesterday, but you already did the job !

<<
    FIXED: mutex conflict

    Mutex was defined as Static, so when more than one SVBONY CMOS camera was connected, one mutex would conflict.
>>

Right, I totally missed this point, thanks !

<<
    IMPROVED: OFFSET support on Ekos Camera module.

    Changed the name of the input field from CCD_DOFFSET to CCD_OFFSET to enable the OFFSET field in the Capture Module of Ekos.
>>

Right again, no need to use a custom value there. Thanks !

I just noticed some minor issues :

- Log messages don't show the target temperature, but the current one (cf. joined screenshot) :

	[INFO] Setting CCD temperature to +03.00 C
	
- I was not able to set a target temperature bellow zero °C

- I don't think the temperature ramping is working

Let me know if you want to fix these issues by yourself, or if you want me to help.

Every thing else seems to be OK.

Could you also please update these files ? :

- /indi-3rdparty/debian/indi-sv305/changelog
- /indi-3rdparty/indi-sv305/AUTHORS (add you name !)
- /indi-3rdparty/indi-sv305/README
- driver version in the CMakeLists.txt (adding temperature regulation is a major step, in my mind, this driver should be 1.3.0)
- /indi-3rdparty/indi-sv305/doc/sv305-doc.md

Thanks a lot for your work !

Blaise


Screenshot from 2022-06-12 14-12-43

**Comment by thx8411** [comment]
Sorry, I think i did a mistake about these two issues :

<<
    Log messages don't show the target temperature, but the current one (cf. joined screenshot) :

    [INFO] Setting CCD temperature to +03.00 C

    I don't think the temperature ramping is working
>>

I need more testing.

Best regards,

Blaise

**Comment by jctk** [comment]
Hi, @thx8411 
Thanks for your review.

Mutex
I detected some more my problem about mutex. I'm correcting it now.

> Let me know if you want to fix these issues by yourself, or if you want me to help.
OKay, I will fix those issues later.

> I was not able to set a target temperature bellow zero °C
Can you tell me how to reproduce the issue where I can't set below ZERO?
I am able to set it in my environment.

> I don't think the temperature ramping is working
You are right.
This is not yet supported.

> Could you also please update these files ? :
Okay, I understand. Update those files later.

And more.
I found an important problem.
It seems that data exceeding 12 bits is truncated to 0. 

SV405CC and SV905C accept following Video format.
 SVB_IMG_RAW8, SVB_IMG_RAW16, SVB_IMG_Y8, SVB_IMG_RGB24
This fact can be seen in Sv305CD.cameraProperty.SupportedVideoFormat. 
The current version of indi_sv305_ccd uses SVB_IMG_RAW12 instead of SVB_IMG_RAW16. 

Can I change from SVB_IMG_RAW12 to SVB_IMG_RAW16 and remove selector of '12bit 16bit stretch' ?
Can I change the color SV305 series in the same way too?
I don't own an SV305 so I don't know if the SV305 has the same problem. 

Best regards,
Kakura.

**Comment by jctk** [comment]
Hi, @thx8411 

I added to following files to the Pull Request.

/indi-3rdparty/debian/indi-sv305/changelog
/indi-3rdparty/indi-sv305/AUTHORS
/indi-3rdparty/indi-sv305/README
//indi-3rdparty/indi-sv305/CMakeLists.txt
/indi-3rdparty/indi-sv305/doc/sv305-doc.md

Could you review the files ?

Best regards,
Kakura.

**Comment by jctk** [comment]
Hi, @thx8411

I have checked the following issues you pointed out.

1. Log messages don't show the target temperature, but the current one.

- Log message is not displayed when current and target temperatures are equal. This is the same behavior as in the generic_ccd.cpp and is not a defect.

2. I was not able to set a target temperature bellow zero °C

- I can set it below zero °C。

3. I don't think the temperature ramping is working

- I would like to know how temperature ramping works. I am new to using a cooling camera and do not understand the desired behavior.
- Does this mean that the temperature changes every minute according to the ramp set in the 'Set Temperature regulation' dialog? If this understanding is correct, I do not want to implement it because SVBONY's API does not allow such control.

Best regards,
Kakura.

**Comment by jctk** [comment]
Hi, @thx8411 

I commited following BUGFIX today.
I believe I have made all the necessary corrections.

Could you review the files and confirm the operation ?
Operation has been confirmed for SV405CC and SV905C.
Could you confirm the operation of your SV305 series?

#BUGFIX
1. Fix some mutex processing
    - Comment out unnecessary pthread_mutex_unlock in Sv305::Disconnect.
    - Added mutex processing in Sv305CCD::SetTemperature.
    - Added mutex processing for getting temperature in Sv305CCD::TimerHit.
    - Fixed from pthread_mutex_unlock to pthread_mutex_lock in Sv305CCD::updateControl
    - Added pthread_mutex_unlock in an error condition in Sv305CCD::GuideNorth, GuideSouth, GuideEast and GuideWest

2. Added 'break' in Sv305CCD::ISNewSwitch

3. Fix generation of luminance 0 data by specifying SVB_IMG_RAW12
    - Changed Frame Format from Raw 12 bit to Raw 16 bit.
    - Specifying SVB_IMG_RAW12 would round values exceeding 12 bits to generate data with luminance 0.
    - This fix is just a HOTFIX. Originally the Frame Format switch should be dynamically generated according to the camera properties.
    - NOTE: The SV405CC and SV905C have been tested on actual equipment with this modification; the SV305 series needs to be tested.

File: debian/indi-sv305/changelog [hunk 1]
```diff
@@ -1,3 +1,9 @@
+indi-sv305 (1.3.0) kakura; urgency=low
+
+  * SV405CC support added
+  * Mutex conflicts fix
+  * OFFSET support on Ekos Camera module
+
```
Comment (thx8411) [inline debian/indi-sv305/changelog:6]: You must add this line "-- Blaise-Florentin Collin <thx8411@yahoo.fr>  Thu, 24 Feb 2022 16:00:00 +0300", with updated author and date, or the package will be broken.

Debain the package changelog file is very strict. See https://www.debian.org/doc/debian-policy/ch-source.html#s-dpkgchangelog

**Comment by thx8411** [comment]
Hi jctk,

Sorry, I'm very busy these days.
I will review your PR further ASAP.

Best regards,

Thx8411

**Comment by jctk** [comment]
I have modified Changelog.
Thank you for letting me know the web of description rules of Debian Changelog.

**Comment by knro** [comment]
@thx8411 Are there any more pending issues?

**Review by thx8411 — APPROVED** [review]
Hi,

Thanks for your help !

I still get some weird issues about temp regulation

Example :

I just plugged the device and started the driver
Camera was 27°C (pretty hot days in France...)
I set 22°C
Driver log : "2022-06-18T08:05:30: [INFO] Setting CCD temperature to +24.50 C"
Actual camera temperature after a few minutes : 8.5 °C

But everything else is OK
Code is clean
Tested on SV305 too -> no regression

I think we can merge "as is".
And fix the temperature regulation later.

Thanks !

Thx8411

**Comment by jctk** [comment]
Hi, @thx8411 
Thanks for the review.

I am also aware of some cooler related issues.
I will fix them after the merge.

Best regards,
Kakura.

**Comment by thx8411** [comment]
Hi jctk,

Do you mind to fork from my repository for you future works ?
It will be easier to work together.

My repo is up to date.

Best regards !

**Comment by jctk** [comment]
Hi @thx8411,
II agree with your proposal.
I will fork from your repository.

I will PR the following two things in the near future.
- Fix timerleft of TimerHit from long to double.
- Simplification of SVBGetVideoData call mechanism

P.S.
SVBONY Camera SDK 1.7.3 has been released.
With this version, RPi frequently freezes on API calls. Please be careful when working with this version.

Best regars,

**Comment by thx8411** [comment]
Thanks !

Thanks also for the advice about the last SDK.

I'm pretty conservative concerning IT updates too ;-)

Best regards !

Blaise

**Comment by thx8411** [comment]
Hi Kakura,

The SDK is not well documented.
I wanted to go a bit deeper in the API behavior.

I wrote a small tool, based on the SVBony demo code base.
So we can  test each "temperature functionality" and see how it works precisely.

Setting the "cooler power : 0% - 100%" seems broken in the firmware (vers 1.6.5).

 Best regards !

Blaise


[SVBDemo.cpp.txt](https://github.com/indilib/indi-3rdparty/files/8935387/SVBDemo.cpp.txt)

**Comment by jctk** [comment]
I have tried to add Cooler Power to the INDI control panel.
At that time, I was using SDK 1.6.1 and could not change the Power; I assumed that Cooler Power's API was not writable.
I don't know if that guess is correct, because I didn't investigate further.
Also, the SDK was not well documented.

We hope that all problems will be corrected in the revised version of SDK 1.7.3.

**Comment by thx8411** [comment]
You're right, Cooler Power is read only, and managed by the camera.


[SVBDemo.cpp.txt](https://github.com/indilib/indi-3rdparty/files/8935920/SVBDemo.cpp.txt)

**Comment by jctk** [comment]
Hi, Blaise,

Indicate Cooler Power:
Valerio Faiuolo, in the Facebook group "SVBONY Telescope Club", provided me with the following patch that shows Cooler Power in the INDI control panel.
https://github.com/olosnet/sv305_ccd

It seems to have all the elements to incorporate there.
I will incorporate it in due course.


Temperature display issue:
This problem stems from the fact that I developed it as I had it in my head, without a clear specification.

I am in the process of breaking down the case for events such as operations from Ekos and INDI.
After that, I plan to check the pattern of method calls, including the source code of the CCD class, in the debugger to define the behavior that should be there.

I intend to take my time and modify it carefully.

Best regards
Kakura

**Comment by thx8411** [comment]
Hi !

My SDK version was the main issue. Switched to 1.6.5

Does any body have the SDK 1.6.5 for MacOS ? We can only download the last one on the SVBony site.

Pushed a new version

Also changed :

- no need to read temp in the SetTemperature method
- add power cooling status (0%-100%) in UI
- set a default target temperature (25 deg) at start
- TemperatureRequest init missing
- copy/paste error on CoolS switch change
- uses Ramping temp Threshold instead of magic number

Still an issue on ramping.
Everything else seems ok.

Best regards !

Blaise

**Comment by jctk** [comment]
Hi Blaise

You fixed it in the middle of the night in Japan while I was sleeping.
Then I will leave the temperature issue to you.

I have no idea about SDK 1.6.5 for Mac.
You may want to contact SVBONY.

Thanks.
Kakura

**Comment by jctk** [comment]
Hi Blaise
I found SDK 1.6.5 for MacOS at SVBONY Japanese site.
https://www.svbony.jp/downloads
There are still SDKs 1.6.5 for all OS.

**Comment by jctk** [comment]
Hi Blaise

Good thing there is a Mac SDK.

One more thing.
Last night I made a pull request to your repository.
Could you please review that as well?

Thanks
Kakura

**Comment by thx8411** [comment]
Hi !

I found the issue about temperature ramping.

Parent class INDI:CCD already manages the TemperatureNP status (IPS_OK and IPS_BUSY). If we update TemperatureNP.s in the SV305 driver with IPS_OK, temperature ramping assumes everything is OK, and don't go further.

So, we should remove this in the TimerHit method :

<<
            if (fabs(TemperatureRequest - TemperatureN[0].value) <= TemperatureRampNP[RAMP_THRESHOLD].value) {
                TemperatureNP.s = IPS_OK;
            } else {
                TemperatureNP.s = IPS_BUSY;
            }
>>

Best regards !

Blaise

**Comment by jctk** [comment]
Hi Blaise

As you point out, TimerHit doesn't need that code.
I removed that code from TimerHit and verified that it works fine in debugger. 

I will commit to your repository PR after the fix.

**Pull request merged**